### PR TITLE
Fix creation of multiple KafkaTopics

### PR DESCRIPTION
**Pull request by bartam1**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | yes |
| New feature?    | no |
| API breaks?     | no |
| Deprecations?   | no |
| Related tickets | #747  |
| License         | Apache 2.0 |

### What's in this PR?
It fixes a bug that happens when trying to add multiple KafkaTopic into the kafka cluster.

### Why?
Validating webhook rejects proper KafkaTopic resource.

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)

### CODE REVIEW

File: pkg/webhook/topic_validator.go [hunk 1]
```diff
@@ -154,16 +153,9 @@ func (s *webhookServer) checkExistingKafkaTopicCRs(ctx context.Context,
...
 	kafkaTopicList := banzaicloudv1alpha1.KafkaTopicList{}
-	err := s.client.List(ctx, &kafkaTopicList,
-		client.MatchingFieldsSelector{
-			Selector: fields.OneTermEqualSelector("spec.name", topic.Spec.Name),
-		},
-		client.MatchingFieldsSelector{
-			Selector: fields.OneTermEqualSelector("spec.clusterRef.name", topic.Spec.ClusterRef.Name),
-		},
-	)
+	err := s.client.List(ctx, &kafkaTopicList, client.MatchingFields{"spec.name": topic.Spec.Name})
```
Comment (stoader) [inline pkg/webhook/topic_validator.go:156]: Can't we add the `spec.clusterRef.name` to the list of fields to match by?

e.g.:
```
client.MatchingFields{"spec.name": topic.Spec.Name, "spec.clusterRef.name", topic.Spec.ClusterRef.Name}
```

File: pkg/webhook/topic_validator.go [hunk 2]
```diff
@@ -154,16 +153,9 @@ func (s *webhookServer) checkExistingKafkaTopicCRs(ctx context.Context,
...
 	if err != nil {
-		log.Info("failed to list KafkaTopics")
+		log.Info(fmt.Sprintf("Failed to list KafkaTopics. Error message: %s", err))
```
Comment (stoader) [inline pkg/webhook/topic_validator.go:158]:
```suggestion
		log.Error(err, "couldn't list KafkaTopic custom resources")
```

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/webhook/topic_validator.go [hunk 3]
```diff
@@ -175,9 +167,12 @@ func (s *webhookServer) checkExistingKafkaTopicCRs(ctx context.Context,
 		}
 
 		referredNamespace := kafkaTopic.Spec.ClusterRef.Namespace
-		if (kafkaTopic.GetNamespace() == clusterNamespace && referredNamespace == "") || referredNamespace == clusterNamespace {
-			foundKafkaTopic = &kafkaTopicList.Items[i]
-			break
+		referredName := kafkaTopic.Spec.ClusterRef.Name
+		if referredName == topic.Spec.ClusterRef.Name {
```
Comment (stoader) [inline pkg/webhook/topic_validator.go:171]: This `if` statement will always evaluate to `true`

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/webhook/topic_validator.go (hunk 1 above)
Comment (bartam1) [reply pkg/webhook/topic_validator.go:156]: No we cant :( This error message happen when I tried this solution: Failed to list KafkaTopics. Error message: non-exact field matches are not supported by the cache"}

File: pkg/webhook/topic_validator.go (hunk 3 above)
Comment (bartam1) [reply pkg/webhook/topic_validator.go:171]: Why? Its kafkaTopic.Spec.ClusterRef.Name == topic.Spec.ClusterRef.Name

Comment (stoader) [reply pkg/webhook/topic_validator.go:171]: I see maybe we could change the name as it's quite easy to confuse.

File: pkg/webhook/topic_validator.go (hunk 1 above)
Comment (hi-im-aren) [reply pkg/webhook/topic_validator.go:156]: For that to work we have to index the fields in the client cache e.g.
```
func addIndexerForKafkaTopic(ctx context.Context, newCache cache.Cache) error {
	nameIndexFunc := func(obj client.Object) []string {
		return []string{obj.(*v1alpha1.KafkaTopic).Spec.ClusterRef.Name}
	}
	err := newCache.IndexField(ctx, &v1alpha1.KafkaTopic{}, "spec.clusterRef.name", nameIndexFunc)
	if err != nil {
		return errors.WrapIfWithDetails(err, "could not setup indexer for field", "field", "spec.clusterRef.name")
	}

	namespaceIndexFunc := func(obj client.Object) []string {
		return []string{obj.(*v1alpha1.KafkaTopic).Spec.ClusterRef.Namespace}
	}
	err = newCache.IndexField(ctx, &v1alpha1.KafkaTopic{}, "spec.clusterRef.namespace", namespaceIndexFunc)
	if err != nil {
		return errors.WrapIfWithDetails(err, "could not setup indexer for field", "field", "spec.clusterRef.namespace")
	}
	return nil
}
```

Comment (bartam1) [reply pkg/webhook/topic_validator.go:156]: Multiple field selection is not supported. https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/internal/cache_reader.go#L119

Comment (stoader) [reply pkg/webhook/topic_validator.go:156]: @hi-im-aren I believe that's already used. The k8s `client` used by the topic validator is the same used by the controller-runtime: https://github.com/banzaicloud/koperator/blob/master/pkg/webhook/server.go#L63

https://github.com/banzaicloud/koperator/blob/master/main.go#L111

Not sure if this is related:
https://github.com/kubernetes-sigs/controller-runtime/blob/7ba3e559790c5e3543002a0d9670b4cef3ccf743/pkg/cache/internal/cache_reader.go#L192

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/internal/cache_reader.go#L119
but this looks as though only matching by a single field is supported.

File: pkg/webhook/topic_validator.go (hunk 3 above)
Comment (bartam1) [reply pkg/webhook/topic_validator.go:171]: I think it can clearly see because the location of the referredName

Comment (stoader) [inline pkg/webhook/topic_validator.go:171]: The conditions in these two `if` statement could be combined into a single expression:
```
if referredName == topic.Spec.ClusterRef.Name &&
  (kafkaTopic.GetNamespace() == clusterNamespace && referredNamespace == "" || 
  referredNamespace == clusterNamespace) {
```

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/webhook/topic_validator.go (hunk 1 above)
Comment (adamantal) [inline pkg/webhook/topic_validator.go:156]: Why is the fieldselector expression got removed? Is it not working?

**Review by adamantal — CHANGES_REQUESTED** [review]
For me it seems that this PR actually removes the client side filtering and use a Go slice based filtering, which essentially does not change the implementation.
I'd like to understand why the client side filtering does not work - and fix it, if it's possible.
Before you start working on this change I'd recommend getting through the original commit: https://github.com/banzaicloud/koperator/commit/dc5c2b9d81a20228a970ed18bc2656f415d13e8f
If we decide to completely remove the client side filtering which has been introduced in that commit, then we also have to:
- delete pkg/k8sutil/cache.go as the `AddKafkaTopicIndexers` is no longer needed
- remove the `AddKafkaTopicIndexers` call from main.go.
- rewrite the implementation that uses Go slice based filtering (already done in this PR)
We might as well just simply revert #717 for sake of simplicity, if it cannot be remediated.

I am open to your thoughts, let's discuss this.

File: pkg/webhook/topic_validator.go (hunk 1 above)
Comment (Kuvesz) [reply pkg/webhook/topic_validator.go:156]: It seemed to be what caused the problem. Strangely enough if only the first selector was in it worked but not with both of them I think.

File: pkg/webhook/topic_validator.go (hunk 3 above)
Comment (bartam1) [reply pkg/webhook/topic_validator.go:171]: Yes it can combined but for me it looks better to separate

File: pkg/webhook/topic_validator.go (hunk 1 above)
Comment (stoader) [reply pkg/webhook/topic_validator.go:156]: Apparently the `client.List` takes into account only the first term from the matching field selector under the hood:
https://github.com/kubernetes-sigs/controller-runtime/blob/7ba3e559790c5e3543002a0d9670b4cef3ccf743/pkg/cache/internal/cache_reader.go#L195
https://github.com/kubernetes-sigs/controller-runtime/blob/7ba3e559790c5e3543002a0d9670b4cef3ccf743/pkg/cache/internal/cache_reader.go#L126

**Review by stoader — APPROVED** [review]

**Review by adamantal — APPROVED** [review]
We discussed this in more detail offline.
TLDR: the indexers are working, but the Kubernetes Go client can only take into account one of this option. It's ok to keep the indexer based solution, but can only use one condition.

**Review by baluchicken — APPROVED** [review]
